src/request.rs

```diff
@@ -368,32 +453,90 @@ impl Request {
+    ///
+    /// The url is wrapped in a `Result` since a common use case is to construct
+    /// the [`Request`] using a `&str` in which case the url is parsed.
     ///
     /// ```
     /// # fn main() -> Result<(), ureq::Error> {
     /// # ureq::is_test(true);
     /// let req = ureq::get("http://httpbin.org/get")
-    ///     .query("foo", "42")
-    ///     .query("foo", "43");
+    ///     .query("foo", "bar");
     ///
-    /// assert_eq!(req.query_params(), vec![
-    ///     ("foo".to_string(), "42".to_string()),
-    ///     ("foo".to_string(), "43".to_string())
-    /// ]);
+    /// assert_eq!(req.parsed_url().unwrap().host(), "httpbin.org");
     /// # Ok(())
     /// # }
     /// ```
-    pub fn query_params(&self) -> Vec<(String, String)> {
-        let mut ret = vec![];
+    pub fn parsed_url(&self) -> Result<&ParsedUrl> {
+        // If there is a parse error, surface it now.
+        let _ = self.parsed_url.result.as_ref()?;
 
-        if let Ok(url) = &self.parsed_url.0 {
-            for (k, v) in url.query_pairs() {
-                ret.push((k.into(), v.into()));
-            }
+        return Ok(&self.parsed_url);
+    }
+
+    /// Get the url str that will be used for this request.
+    ///
+    /// The url might differ from that originally provided when constructing the
+    /// request if additional query parameters have been added using [`Request::query()`].
+    ///
+    /// In case the original url provided to build the request is not possible to
+    /// parse to a Url, this function returns the original, and it will error once the
+    /// Request object is used.
+    ///
+    /// ```
+    /// # fn main() -> Result<(), ureq::Error> {
+    /// # ureq::is_test(true);
+    /// let req = ureq::get("http://httpbin.org/get")
+    ///     .query("foo", "bar");
+    ///
+    /// assert_eq!(req.url(), "http://httpbin.org/get?foo=bar");
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// ```
+    /// # fn main() -> Result<(), ureq::Error> {
+    /// # ureq::is_test(true);
+    /// let req = ureq::get("SO WRONG")
+    ///     .query("foo", "bar"); // does nothing
+    ///
+    /// assert_eq!(req.url(), "SO WRONG");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn url(&self) -> &str {
+        if let Ok(url) = &self.parsed_url.result {
+            url.as_str()
+        } else {
+            self.parsed_url
+                .origin
+                .as_ref()
+                .map(|s| s.as_str())
+                // This should not happen since we either created Request using a valid Url
+                // object, in which case we have Ok(url) above, or we used a String, in
+                // which case we definitely have a value in self.parsed_url.1
+                .expect("Illegal URL without original")
         }
+    }
 
-        ret
+    /// Test whether the url for this request is valid. If this test is
+    /// `false`, the resulting use of this Request will fail with an error.
```

Comment:
This feels like an awkward function to use. What's the expected use case?

Reply:
Let's remove it.

Reply:
Gone.